x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/QueryBuilderResolver.java

```diff
@@ -0,0 +1,170 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.session;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ResolvedIndices;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryRewriteContext;
+import org.elasticsearch.index.query.Rewriteable;
+import org.elasticsearch.search.SearchService;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.esql.core.util.Holder;
+import org.elasticsearch.xpack.esql.expression.function.fulltext.FullTextFunction;
+import org.elasticsearch.xpack.esql.plan.logical.EsRelation;
+import org.elasticsearch.xpack.esql.plan.logical.LogicalPlan;
+import org.elasticsearch.xpack.esql.planner.PlannerUtils;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiConsumer;
+
+/**
+ * Some {@link FullTextFunction} implementations such as {@link org.elasticsearch.xpack.esql.expression.function.fulltext.Match}
+ * will be translated to a {@link QueryBuilder} that require a rewrite phase on the coordinator.
+ * {@link QueryBuilderResolver#resolveQueryBuilders(LogicalPlan, ActionListener, BiConsumer)} will rewrite the plan by replacing
+ * {@link FullTextFunction} expression with new ones that hold rewritten {@link QueryBuilder}s.
+ */
+public class QueryBuilderResolver {
+    private final SearchService searchService;
+    private final ClusterService clusterService;
+    private final TransportService transportService;
+    private final IndexNameExpressionResolver indexNameExpressionResolver;
+
+    public QueryBuilderResolver(
+        SearchService searchService,
+        ClusterService clusterService,
+        TransportService transportService,
+        IndexNameExpressionResolver indexNameExpressionResolver
+    ) {
+        this.searchService = searchService;
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        this.indexNameExpressionResolver = indexNameExpressionResolver;
+    }
+
+    public void resolveQueryBuilders(
+        LogicalPlan plan,
+        ActionListener<Result> listener,
+        BiConsumer<LogicalPlan, ActionListener<Result>> callback
+    ) {
+        if (plan.optimized() == false) {
+            throw new IllegalStateException("Expected optimized plan before query builder rewrite.");
+        }
+
+        Set<FullTextFunction> unresolved = fullTextFunctions(plan);
+        Set<String> indexNames = indexNames(plan);
+
+        if (indexNames == null || indexNames.isEmpty() || unresolved.isEmpty()) {
+            callback.accept(plan, listener);
+            return;
+        }
+        QueryRewriteContext ctx = queryRewriteContext(indexNames);
```

Comment:
I wonder if `QueryRewriteContext` an intermediate data structure to carry `ResolvedIndices` to get the inference results? Can we just provide `ResolvedIndices` instead, without having to create a `QueryRewriteContext`?